src/parse.py

```diff
@@ -0,0 +1,285 @@
+        curried_throw(NotImplementedError(f"Invalid model type! Got: `{data['type']}`"))
+    )
+
+
+    del data['type']
+
+    try:
+        return model_class(**data)
+    except Exception as e:
+        print(f"Unexpected error when instantiating model!: \n\t{e}")
+
+
+def get_function_class(x_dist: D.Distribution, x_curr_dim: int, data: dict) -> Optional[FunctionClass]:
+    if 'type' not in data:
+        raise KeyError(f"Function Class type not specified!")
+
+    f_class_type: type[FunctionClass] | Callable[[], None] = FUNCTION_CLASSES.get(data['type'], 
+        curried_throw(NotImplementedError(f"Invalid function class! Got: `{data['type']}`"))
+    )
+
+    data.update({
+        "x_distribution" : x_dist,
+        "x_curriculum_dim" : x_curr_dim,
+    })
+
+    del data['type']
+    try:
+        return f_class_type(**data)
+    except Exception as e:
+        print(f"Unexpected error when instantiating model!: \n\t{e}")
+
+
+def get_optimizer(model: ContextModel, data: dict) -> Optional[torch.optim.Optimizer]:
+    if 'type' not in data:
+        raise KeyError(f"Optimizer type not specified!")
+
+    OPTIMIZERS = {
+        "sgd" : torch.optim.SGD,
+        "adam": torch.optim.Adam
+    }
+
+    optim_type: type[torch.optim.Optimizer] | Callable[[Any], None] = OPTIMIZERS.get(data['type'],
+        curried_throw(NotImplementedError(f"Invalid optimizer! Got: `{data['type']}`"))
+    )
+
+    del data['type']
+    try:
+        return optim_type(model.parameters(), **data)
+    except Exception as e:
+        print(f"Unexpected error when instantiating optimizer!: \n\t{e}")
+
+
+def get_loss_fn(data: dict) -> Optional[torch.nn.Module]:
+    if 'type' not in data:
+        raise KeyError(f"Loss function type not specified!")
+    
+    LOSS_FNS = {
+        "squared" : torch.nn.MSELoss
+    }
+
+    loss_fn_type: type[torch.nn.Module] | Callable[[], None] = LOSS_FNS.get(data['type'],
+        curried_throw(NotImplementedError(f"Invalid option for loss function! Got: {data['type']}"))
+    )
+
+    del data['type']
+    try:
+        return loss_fn_type(**data)
+    except Exception as e:
+        print(f"Unexpected error when instantiating loss function!: \n\t{e}")
+
+
+def produce_trainer_stages(data: dict) -> list[ContextTrainer]:
+    """Convert a list of YAML primitive stage dicts to a list of dictionaries with instantiated objects"""
+
+    x_dim: int = _get_value(data['x_dim'], int(1e99)) 
+    stages, step_counts = expand_curriculum(data)
+    for i in range(len(stages)):
+        stages[i]['steps'] = step_counts[i]
+    
+    for stage in stages:
+        b_size  = stage['b_size']
+        seq_len = stage['seq_len']
+        x_curriculum_dim = stage['x_dim']
+
+        stage['x_dist'] = get_x_distribution(
+            b_size, seq_len, x_dim, stage.get('x_dist', {})
+        )
+
+        stage['model'] = get_model(
+            stage['model'] | { "x_dim" : x_dim }
+        )
```

Comment:
I could be wrong, but this seems to create a new model for each stage since `get_model` creates a new model (of the same type, but not shared parameters) each time its called.

potential solution:
- add: `model = None` before the for loops
- add after the current line 235: `model = model or stage['model']`
- add after adding the previous line: `stage['model'] = model`

Reply:
Nice catch! I think making one model outside the loop and referencing it for each stage would be the same:
```python
model = get_model(stages[0]['model'] | { "x_dim" : x_dim })
for stage in stages:
    ...
    stage['model'] = model
    ...
```

The `or` in the first solution proposed might be a bit verbose, since a curriculum necessarily must have a linear ancestry of models before it (so is always equivalent to just one model shared for all stages)